### ISSUE TITLE
Test coverage of message specifications

### ISSUE THREAD

**Issue by treiher** [Componolit/RecordFlux-specifications#11]
It would be helpful to determine the test coverage of the message specifications. The validation tool (#7) would need to analyze how many paths of the message graph are covered by the existing test data.

----
Implementation Plan

MessageValue
- [x] Add attribute (property) `covered_links: List[Link]`
- [x] Add `_next_link(field_name) -> Link` method which returns the next valid link 
- [x] Call `next_link` inside `next_field` and store the last returned link in `covered_links`
- [x] Add `get_sdu_messages -> List[MessageValue]` which returns the list of all messages contained in all fields of type `OpaqueValue` and `ArrayValue` 
- [x] Add tests

Validation Tool

- [x] create dict to store the coverage of all involved messages (models) - total coverage
- [x] method to update the total coverage on every iteration
- [x] generate output
- [x] tests

Output

Short Output: 
```
------------------------------------------------------------------------------
                     RecordFlux Validation Coverage Report
Directory: .
------------------------------------------------------------------------------
File                                       Links    Used   Coverage  
------------------------------------------------------------------------------
ethernet.rflx                                100     100   100%
in_ethernet.rflx                              34      34   100%
ipv4.rflx                                    206      97    47%   
------------------------------------------------------------------------------
TOTAL                                        340     231    68%
------------------------------------------------------------------------------
```

Full Output:
`<file_name>:<line>:<col>: error: missing link <link_source_field> -> <link_target_field>: <link_condition>`

**Comment by rssen** [comment]
Output: 
1) print list of links that were not used over the course of the testing session (separated by spec file and message)
2) Percentage of covered links
```
<file_name>:<message_identifier>:
links covered: <percentage>
missed:
   <link_source_field> -> <link_target_field>: <link_condition>
```

**Comment by senier** [comment]
Here is what I imagine (analogous to `gcov`):

```Console
------------------------------------------------------------------------------
                     RecordFlux Validation Coverage Report
Directory: .
------------------------------------------------------------------------------
File                                       Links    Used  Cover   Missing
------------------------------------------------------------------------------
ethernet.rflx                                100     100   100%
in_ethernet.rflx                              34      34   100%
ipv4.rflx                                    206      97    47%   4:67-5:13, 7:5, 22:77-23:18, 114:5-116:64 
------------------------------------------------------------------------------
TOTAL                                        340     231    68%
------------------------------------------------------------------------------
```

Note, that while we work on links internally, the positions in the `Missing` column are the source locations of these very links. Contrary to `gcov` we should add the columns (separated by `:`) to the output.

**Comment by rssen** [comment]
The coverage tool should check if an error link has been hit. The condition of this error link should consist of the negation of the disjunction of the conditions of all outgoing edges.

**Comment by treiher** [comment]
In addition: There might not be an invalid link if there is no condition at the outgoing link or all cases are covered by outgoing links. Type constraints need also to be considered, e.g., the value constraints of a range integer must be considered as a condition at all outgoing links.

**Comment by rssen** [comment]
> Type constraints need also to be considered, e.g., the value constraints of a range integer must be considered as a condition at all outgoing links.

Consequently, we would also have to split up `or` conditions and create new links for every possible enum value for a field of type `Enum`. I have a strong feeling that the creation of invalid paths is closely related to what the fuzzer did in its original version.

**Comment by treiher** [comment]
It depends on our goal. In the fuzzer we wanted to hit multiple invalid values (all variations with a high probability for bugs), so we had to split up some conditions into multiple links. That could be also interesting for the coverage analysis at some point. For example this could be used to validate the correctness of the fuzzer, i.e., check that the fuzzer really looks for all interesting variations. I think for the beginning we should keep it simple and just adding one link covering all invalid values. We can make the analysis more sophisticated at a later stage.

**Comment by senier** [comment]
> We can make the analysis more sophisticated at a later stage.

When looking at HTTP/2, I think we really want to split disjunctions (cf. #31), but I agree that we should build the simple version first and extend it later.

**Comment by rssen** [comment]
I updated the first comment with the implementation plan. (Not sure if you got a notification for that)

**Comment by treiher** [comment]
> I updated the first comment with the implementation plan. (Not sure if you got a notification for that)

Thanks for the info. There is no notification for edited comments. In the following I added some comments about the implementation plan. It probably makes sense to discuss the issues in a call.

> Implementation Plan
> 
> Field (Model)
> I know we discussed the option of inserting two boolean variables into a field:
> 
>     * Add boolean variable `incoming_error_link`, that signals, if the previous field has an outgoing error link.
>       The prev field does not have an error link, if the disjunction of the conditions of all outgoing links evaluate to true.
> 
>     * Add boolean variable `incoming_error_link_used`, that signals, if the error link  of the previous field has been used so far.
> 
> 
> Maybe i got it wrong or missed something but i currently see two problems with this:
> 
>     1. How should a situation be handled where a field has two or more predecessors? From my understanding this would require two boolean variables in a field for each predecessor field.
> 
>     2. What if a field had two or more successors? Each successor field would need to be updated once the error edge is used (an exception is raised)

I don't think we want to have such boolean variables in a field. Also the outgoing links need to be analyzed to determine if there is an error link, not the incoming links. An error link exists if there is a field value which is not covered by a condition at any outgoing link.

> For now, i would favour the following solution:
> 
> Link (Model)
> 
>     * [ ]  Add boolean variable `covered` that signals if the link was used
> 
>     * [ ]  Initialize to False

The model should not be changed for such a very specific functionality. The information if a link is covered should be hold in the coverage tool.

> MessageValue
> 
>     * [ ]  Add coverage flag
> 
>     * [ ]  While parsing a binary message, set the `covered` variable to `true` in the `next_field` method for the link that is used to get to the next field
> 
>     * [ ]  Add a method to clear the state of the MessageValue (maybe already present), model stays untouched

It is unclear to me, why there needs to be a coverage flag in `MessageValue`.

> Validation Tool
> 
> [...]
>
>     * [ ]  Insert error links into the model
>       
>       * Error link will be taken, if a parsed value is invalid and can be set to `true`
>
> [...]
>
> This would offload the heavy lifting of creating error links and deciding if an error link should be created in the first place into the validation tool instead of MessageValue or Model. As we discussed splitting up conditions with disjunctions in the future, where new edges also would need to be inserted into the model, this would prepare it and reduce the workload for this feature.

Is it strictly necessary to add error links to the model? Modifying the model could interfere with the message validation. I think it would make sense to do the validation and the coverage analysis at the same time. Do you have an idea how to do that?
 
> Needs to be solved:
> 
> [...]
>       
>       * After a field has been parsed, the next field is determined by the next_field method but the value of the next field is not known at this point. An edge (A) could be considered as covered only after its target field (T) has been set successfully. That would require to store (A) in a variable until the `next_field` method is called again for (T) after its field value has been set.

I don't understand the problem. Link conditions only depend on previous fields. If the source of a link and all predecessors are set, it should be clear which link is valid.

>     * Is the `simplified` method capable of simplifying a complex condition to `TRUE` or `FALSE`? I remember this has been a problem in the past.

The clear answer is: It depends.

If all field values are known, it should be able to completely simplify the expression.

**Comment by treiher** [comment]
Extended output:

```
<file_name>:line:col: error: missing link <link_source_field> -> <link_target_field>: <link_condition>   
```

**Comment by rssen** [comment]
```
------------------------------------------------------------------------------
                     RecordFlux Validation Coverage Report
Directory: .
------------------------------------------------------------------------------
File                                       Links    Used   Coverage  
------------------------------------------------------------------------------
ethernet.rflx                                100     100   100%
in_ethernet.rflx                              34      34   100%
ipv4.rflx                                    206      97    47%   
------------------------------------------------------------------------------
TOTAL                                        340     231    68%
------------------------------------------------------------------------------
```
Do we really want the coverage information on a per-file basis? 
I would think that it is more useful to know the coverage on a per-message basis. There could be messages in a file that are not used (and not needed) during a test session.

**Comment by senier** [comment]
> Do we really want the coverage information on a per-file basis?
> I would think that it is more useful to know the coverage on a per-message basis. There could be messages in a file that are not used (and not needed) during a test session.

IMHO our first attempt should be as consistent with `gcov` as possible. When analyzing the coverage of say a C file, there will also be functions that are not covered. If so, you either have to write additional tests or annotated the code (`gcov` as special comments for this, we don't). Our goal should be to end up with full coverage of all messages in a spec.

> And what should be the `Directory`? The current working directory or the directories containing the test cases?

In `gcov` this is the current working directory (which must be the same as the one the compiler was called in). Let's use that and see whether it is useful in our case.

**Comment by rssen** [comment]
Regarding the extended output:
I noticed that there are many links without location information. This makes it hard to see where a certain link, that is not covered yet, is in the spec file (one could guess from the source and target field but that is not ideal).  Is that a known problem?
An excerpt:

```
--------------------------------------------------------------------------------
                     RecordFlux Validation Coverage Report                      
Directory: /home/rs/Development/PYTHON/PycharmProjects/RecordFlux-specifications
--------------------------------------------------------------------------------
File                                          Links       Used        Coverage
ipv4.rflx                                        33         18          54.55%
ethernet.rflx                                    12         12         100.00%
arp.rflx                                         10          0           0.00%
ipv6.rflx                                       105         18          17.14%
--------------------------------------------------------------------------------
TOTAL                                           160         48          30.00%
--------------------------------------------------------------------------------


================================================================================
                                 Link Coverage                                  
================================================================================


                                   ipv4.rflx                                    
--------------------------------------------------------------------------------
None: error: missing link Initial -> Option_Type_Action: True
ipv6.rflx:63:13: error: missing link Opt_Data_Len -> Option_Data: True
None: error: missing link Option_Data -> Final: True
None: error: missing link Option_Type_Action -> Option_Type_May_Change: True
None: error: missing link Option_Type_May_Change -> Option_Type_Value: True
ipv6.rflx:54:13: error: missing link Option_Type_Value -> Final: Option_Type_Value = Pad1
and Option_Type_Action = Skip
and Option_Type_May_Change = False
ipv6.rflx:58:13: error: missing link Option_Type_Value -> Opt_Data_Len: Option_Type_Value /= Pad1
or Option_Type_Action /= Skip
or Option_Type_May_Change = True
ipv6.rflx:280:13: error: missing link Destination_Data -> Payload: True
None: error: missing link Flow_Label -> Payload_Length: True
None: error: missing link Hop_Limit -> Source_Data: True
None: error: missing link Next_Header -> Hop_Limit: True
None: error: missing link Payload_Length -> Next_Header: True
None: error: missing link Source_Data -> Destination_Data: True
None: error: missing link Traffic_Class -> Flow_Label: True
None: error: missing link Version -> Traffic_Class: True
```

**Comment by treiher** [comment]
> I noticed that there are many links without location information. This makes it hard to see where a certain link, that is not covered yet, is in the spec file (one could guess from the source and target field but that is not ideal). Is that a known problem?

I suspect the link location is missing for all links for which there is no direct representation (i.e. a then-clause) in the specification. It should be possible to adapt the parser to use the location of the target field or the source field in these cases.

**Comment by treiher** [comment]
> I suspect the link location is missing for all links for which there is no direct representation (i.e. a then-clause) in the specification. It should be possible to adapt the parser to use the location of the target field or the source field in these cases.

Will be fixed in Componolit/RecordFlux#647.
